datapower-docker-compose-dp-mq/datapower-docker-compose-dp-mq.mkd

```diff
@@ -0,0 +1,168 @@
+15f5ec8580f1: Pull complete
+b262ae65d728: Pull complete
+d2cd040ac8b2: Pull complete
+ebb4e3f1e29f: Pull complete
+75d31cfdd3e1: Pull complete
+Digest: sha256:7f11236a781c8cdf74092593afda1299ad7372e15c72ca8add22df892af1cc6b
+Status: Downloaded newer image for ddavison/soapui:latest
+Creating src_http_1
+Creating src_mq_1
+Creating src_datapower_1
+Creating src_soapui_1
+Creating src_curldriver_1
+Attaching to src_http_1, src_mq_1, src_datapower_1, src_soapui_1, src_curldriver_1
+.....
+.....
+soapui_1      | 172.19.0.6 - - [07/Jan/2017 22:42:39] "POST / HTTP/1.1" 200 -
+curldriver_1  | 22:42:38,819 DEBUG [SoapUIMultiThreadedHttpConnectionManager$SoapUIDefaultClientConnection] Receiving response: HTTP/1.1 200 OK
+curldriver_1  | SUCCESS! Exiting
+src_curldriver_1 exited with code 0
+
+```
+`src_soapui_1` has assertions set up to validate the message being returned. When the assertions are met, a `200 - OK` message is returned to `curldriver_1` as can be seen in the output above.
+
+Depending on the system on which you are deploying, the first message or two may fail to be processed as the environment may still be building. This is easily detectable in the console logs.
+
+After the first successful response is received the curldriver script exits. You may end the test with the docker command ctrl+c, which will exit the application and stop the containers. To rerun the test, run the `docker-compose up` command again.
+
+## Request Flow- Meet "bob"
+Here I'll outline the travels of the request message, `bob`, through the message flow created in the docker-compose environment. 
+
+First from the Dockerfile, the `src_curldiriver` image is configured to run the following shell script which sends a curl command via HTTP to `src_soapui`.
+```
+#!/bin/bash
+# set env vars to modify operation
+# TIMEOUT=<time in seconds> for amount of time to continue to retry
+# CONTINUOUS=true to continuously test, otherwise exits immediately on success
+
+if [ "$TIMEOUT" -a "$TIMEOUT" -gt 0 ]
+then
+  STOPTIME=$(( $(date +%s) + $TIMEOUT ))
+else
+  STOPTIME=9999999999
+fi
+
+RC=1
+
+while [ $(date +%s) -lt $STOPTIME ]
+do
+  if curl --silent --insecure --form "project=@/usr/local/bin/grow-soapui-project.xml" --form "suite=sendRequestGrow" http://soapui:3000 | grep "200 OK"
+  then
+    RC=0
+    if [ "$CONTINUOUS" != "true" ]
+    then
+      echo "SUCCESS! Exiting"
+      exit $RC
+    fi
+  else
+    echo curl error $?
+  fi
+  sleep 5
+done
+```
+The script curls a SoapUI project named `grow-soapui-project.xml`, located in the /usr/local/bin/ directory, to `src_soapui_1`. The script is set to send a request message every five seconds until a successful transaction is completed. Note that the console log may show a few request failures as the application comes up.
+
+`src_soapui_1` accepts the curl request and processes the test defined in the SoapUI project suite. This is where `bob` is created as an XML message and sent to `src_datapower_1`.
```

Comment:
- SoapUI is acting as a proxy. It takes a http request form curl that contains the soapui project and uses the soapui project to construct the request or requests that it then sends to DataPower.
- Add: We don't need to change anything in soapui! The project file is passed in via curl, so the off-the-shelf soapui DockerHub image is all we need.